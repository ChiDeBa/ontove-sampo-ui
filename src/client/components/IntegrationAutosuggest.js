import React from 'react';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import TextField from '@material-ui/core/TextField';
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import SuggestionItem from './SuggestionItem';
import InputAdornment from '@material-ui/core/InputAdornment';
import CircularProgress from '@material-ui/core/CircularProgress';
import IconButton from '@material-ui/core/IconButton';
import SearchIcon from '@material-ui/icons/Search';

function renderInput(inputProps) {
  const { classes, ref, ...other } = inputProps;

  return (
    <TextField
      fullWidth
      InputProps={{
        inputRef: ref,
        classes: {
          input: classes.input,
        },
        ...other,
      }}
    />
  );
}

function renderSuggestion(suggestion, { query, isHighlighted }) {
  return <SuggestionItem suggestion={suggestion} query={query} isHighlighted={isHighlighted} />;
}

function renderSuggestionsContainer(options) {
  const { containerProps, children } = options;

  return (
    <Paper {...containerProps} square>
      {children}
    </Paper>
  );
}

function getSuggestionValue(suggestion) {
  return suggestion;
}

const styles = theme => ({
  container: {
    flexGrow: 0,
    position: 'relative',
    marginTop: theme.spacing.unit * 2,
    // paddingLeft: theme.spacing.unit * 15,
    // paddingRight: theme.spacing.unit * 15,
    marginLeft: 24,
    //marginRight: 'auto',
    width: 280,
  },
  suggestionsContainerOpen: {
    position: 'absolute',
    zIndex: 1500,
    marginTop: theme.spacing.unit,
    left: 0,
    right: 0,
    marginLeft: 'auto',
    marginRight: 'auto',
    width: 300,
  },
  suggestion: {
    display: 'block',
  },
  suggestionsList: {
    margin: 0,
    padding: 0,
    listStyleType: 'none',
    maxHeight: 500,
    overflow: 'auto',
  },
  icon: {
    color: theme.palette.text.secondary,
    //paddingTop: 4,
  },
  spinner: {
    //margin: theme.spacing.unit,
  }
});

const IntegrationAutosuggest = (props) => {

  let autosuggestDOM = React.createRef();

  const handleOnChange = (event, { newValue }) => {
    props.clearSuggestions();
    props.updateQuery(newValue);
  };

  const handleOnSuggestionSelected = () => {
    searchPlaces();
  };

  const handleOnKeyDown = (event) => {
    if (event.key === 'Enter') {
      searchPlaces();
    }
  };

  const handleClickSearchButton = () => {
    searchPlaces();
  };

  const handleMouseDownSearchButton = (event) => {
    event.preventDefault();
  };

  const searchPlaces = () => {
    if (props.search.query.length > 0) {
      console.log('fetching results');
      autosuggestDOM.current.input.blur();
      props.clearResults();
      props.fetchResults();
    }
  };

  // const handleOnBlur = (event, { highlightedSuggestion }) => {
  //   // console.log(event);
  //   // console.log(highlightedSuggestion);
  // };

  const handleOnSuggestionsFetchRequested = ({ value }) => {
    if (props.search.suggestionsQuery != value || props.search.suggestions.length === 0) {
      console.log('fetching suggestions');
      props.fetchSuggestions();
    }
    else {
      console.log('using old suggestions');
    }
  };

  const shouldRenderSuggestions = (value)  => {
    return value.trim().length > 3;
  };

  const handleOnSuggestionsClearRequested = () => {
    //console.log('SuggestionsClearRequested');
    //props.clearSuggestions();
  };

  const { classes } = props;
  //console.log('IntegrationAutosuggest', props.search.suggestions);

  let adornment = null;
  if (props.search.fetchingSuggestions || props.search.fetchingResults) {
    adornment =
    <InputAdornment position="end">
      <IconButton
        aria-label="Search places"
      >
        <CircularProgress size={24} />
      </IconButton>
    </InputAdornment>;
  } else {
    adornment =
    <InputAdornment position="end">
      <IconButton
        aria-label="Search places"
        onClick={handleClickSearchButton}
        onMouseDown={handleMouseDownSearchButton}
      >
        <SearchIcon className={classes.icon} />
      </IconButton>
    </InputAdornment>;
  }

  return (
    <Autosuggest
      theme={{
        container: classes.container,
        suggestionsContainerOpen: classes.suggestionsContainerOpen,
        suggestionsList: classes.suggestionsList,
        suggestion: classes.suggestion,
      }}
      renderInputComponent={renderInput}
      suggestions={props.search.suggestions}
      shouldRenderSuggestions={shouldRenderSuggestions}
      onSuggestionsFetchRequested={handleOnSuggestionsFetchRequested}
      onSuggestionsClearRequested={handleOnSuggestionsClearRequested}
      renderSuggestionsContainer={renderSuggestionsContainer}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      onSuggestionSelected={handleOnSuggestionSelected}
      focusInputOnSuggestionClick={false}
      ref={autosuggestDOM}
      inputProps={{
        classes,
        placeholder: 'Search place names',
        value: props.search.query,
        onChange: handleOnChange,
        onKeyDown: handleOnKeyDown,
        autoFocus: true,
        endAdornment: adornment
      }}
    />
  );
};

IntegrationAutosuggest.propTypes = {
  classes: PropTypes.object.isRequired,
  search: PropTypes.object.isRequired,
  updateQuery: PropTypes.func.isRequired,
  fetchSuggestions: PropTypes.func.isRequired,
  clearSuggestions: PropTypes.func.isRequired,
  fetchResults: PropTypes.func.isRequired,
  clearResults: PropTypes.func.isRequired
};

export default withStyles(styles)(IntegrationAutosuggest);
