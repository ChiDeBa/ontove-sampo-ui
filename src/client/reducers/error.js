import {
  SHOW_ERROR,
  FETCH_PAGINATED_RESULTS_FAILED,
  FETCH_FACET_FAILED
} from '../actions';

export const INITIAL_STATE = {
  id: 0,
  message: {},
};

const error = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case SHOW_ERROR:
    case FETCH_PAGINATED_RESULTS_FAILED:
    case FETCH_FACET_FAILED:
      return {
        ...state,
        id: state.id + 1,
        message: {
          text: action.message.text,
          title: action.message.title
        }
      };
    default:
      return state;
  }
};

export default error;
