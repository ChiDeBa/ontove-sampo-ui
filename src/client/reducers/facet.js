import {
  FETCH_FACET,
  UPDATE_FACET,
  UPDATE_FILTER,
  OPEN_FACET_DIALOG,
  CLOSE_FACET_DIALOG,
} from '../actions';

export const INITIAL_STATE = {
  source: [],
  productionPlace: [],
  author: [],
  language: [],
  sourceIsFetching: false,
  productionPlaceIsFetching: false,
  authorIsFetching: false,
  languageIsFetching: false,
  facetOptions : {
    productionPlace: {
      id: 'productionPlace',
      label: 'Production place',
      //predicate: defined in backend
    },
    author: {
      id: 'author',
      label: 'Author',
      // predicate: defined in backend
    },
    source: {
      id: 'source',
      label: 'Source',
      // predicate: defined in backend
    },
    language: {
      id: 'language',
      label: 'Language',
      // predicate: defined in backend
    }
  },
  facetFilters: {
    productionPlace: new Set(),
    author: new Set(),
    source: new Set(),
    language: new Set(),
  },
  updatedFacet: ''
};

const facet = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case OPEN_FACET_DIALOG:
      return {
        ...state,
        facetDialogOpen: true,
        activeFacet: action.property
      };
    case CLOSE_FACET_DIALOG:
      return { ...state, facetDialogOpen: false };
    case FETCH_FACET:
      return {
        ...state,
        [ `${action.id}IsFetching` ]: true
      };
    case UPDATE_FACET:
      return {
        ...state,
        [ action.id ]: action.facetValues,
        [ `${action.id}IsFetching` ]: false
      };
    case UPDATE_FILTER:
      return updateFilter(state, action);
    default:
      return state;
  }
};

const updateFilter = (state, action) => {
  const { property, value } = action.filter;
  let valueSet = state.facetFilters[property];
  if (valueSet.has(value)) {
    valueSet.delete(value);
  } else {
    valueSet.add(value);
  }
  const newFacetFilters = {
    ...state.facetFilters,
    [ property ] : valueSet
  };
  return {
    ...state,
    facetFilters: newFacetFilters,
    updatedFacet: property
  };
};

export default facet;
