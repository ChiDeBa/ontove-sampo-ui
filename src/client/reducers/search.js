import {
  UPDATE_QUERY,
  TOGGLE_DATASET,
  FETCH_SUGGESTIONS,
  FETCH_RESULTS,
  UPDATE_SUGGESTIONS,
  CLEAR_SUGGESTIONS,
  UPDATE_RESULTS,
  CLEAR_RESULTS
} from '../actions';

export const INITIAL_STATE = {
  query: '',
  datasets: {
    'kotus': {
      'title': 'Institute for the Languages of Finland (Kotus) Digital Names archive',
      'shortTitle': 'DNA',
      'timePeriod': '1900s',
      'selected': false
    },
    'pnr': {
      'title': 'Finnish Geographic Names Registry',
      'shortTitle': 'FGN',
      'timePeriod': 'contemporary',
      'selected': false
    },
    'tgn': {
      'title': 'The Getty Thesaurus of Geographic Names',
      'shortTitle': 'TGN',
      'timePeriod': '?',
      'selected': true
    },
    'warsa_karelian_places': {
      'title': 'Karelian map names',
      'shortTitle': 'KMN',
      'timePeriod': '1922-1944',
      'selected': false
    },
    'warsa_municipalities': {
      'title': 'Finnish WW2 municipalities',
      'shortTitle': 'FWM',
      'timePeriod': '1939-1944',
      'selected': false
    },
  },
  suggestions: [],
  suggestionsQuery: '',
  fetchingSuggestions: false,
  results: [],
  resultsQuery: '',
  fetchingResults: false,
};

const search = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case UPDATE_QUERY:
      return { ...state, query: action.query || '' };
    case TOGGLE_DATASET:
      return {
        ...state,
        suggestions: [],
        results: [],
        datasets: {
          ...state.datasets,
          [action.dataset]: {
            ...state.datasets[action.dataset],
            selected: state.datasets[action.dataset].selected ? false : true
          }
        }
      };
    case FETCH_SUGGESTIONS:
      return { ...state, fetchingSuggestions: true };
    case FETCH_RESULTS:
      return { ...state, fetchingResults: true };
    case CLEAR_SUGGESTIONS:
      return {
        ...state,
        suggestions: [],
        suggestionsQuery: '',
        fetchingSuggestions: false
      };
    case UPDATE_SUGGESTIONS:
      return {
        ...state,
        suggestions: action.suggestions,
        suggestionsQuery: state.query,
        fetchingSuggestions: false
      };
    case CLEAR_RESULTS:
      return {
        ...state,
        results: [],
        resultsQuery: '',
        fetchingResults: false
      };
    case UPDATE_RESULTS:
      return {
        ...state,
        results: action.results,
        resultsQuery: state.query,
        fetchingResults: false
      };
    default:
      return state;
  }
};

export default search;
