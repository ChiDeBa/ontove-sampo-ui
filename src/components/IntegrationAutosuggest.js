import React from 'react';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import TextField from '@material-ui/core/TextField';
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import SuggestionItem from './SuggestionItem';

function renderInput(inputProps) {
  const { classes, ref, ...other } = inputProps;

  return (
    <TextField
      fullWidth
      InputProps={{
        inputRef: ref,
        classes: {
          input: classes.input,
        },
        ...other,
      }}
    />
  );
}

function renderSuggestion(suggestion, { query, isHighlighted }) {
  return <SuggestionItem suggestion={suggestion} query={query} isHighlighted={isHighlighted} />;
}

function renderSuggestionsContainer(options) {
  const { containerProps, children } = options;

  return (
    <Paper {...containerProps} square>
      {children}
    </Paper>
  );
}

function getSuggestionValue(suggestion) {
  return suggestion.label;
}

const styles = theme => ({
  container: {
    flexGrow: 1,
    position: 'relative',
  },
  suggestionsContainerOpen: {
    position: 'absolute',
    zIndex: 1500,
    marginTop: theme.spacing.unit,
    left: 0,
    right: 0,
  },
  suggestion: {
    display: 'block',
  },
  suggestionsList: {
    margin: 0,
    padding: 0,
    listStyleType: 'none',
    maxHeight: 500,
    overflow: 'auto',
  },
  sectionContainer: {
    borderTop: '1px dashed #ccc',
  },
  sectionContainerFirst: {
    borderTop: 0,
  }
});

const IntegrationAutosuggest = (props) => {

  const handleOnChange = (event, { newValue }) => props.updateQuery(newValue);

  const handleOnSuggestionSelected = () => props.fetchResults();

  const handleOnKeyDown = (event) => {
    if (event.key === 'Enter') {
      props.fetchResults();
    }
  };

  const { classes } = props;

  //console.log('IntegrationAutosuggest', props);

  return (
    <Autosuggest
      theme={{
        container: classes.container,
        suggestionsContainerOpen: classes.suggestionsContainerOpen,
        suggestionsList: classes.suggestionsList,
        suggestion: classes.suggestion,
        sectionContainer : classes.sectionContainer,
        sectionContainerFirst : classes.sectionContainerFirst,
      }}
      renderInputComponent={renderInput}
      suggestions={props.search.suggestions}
      onSuggestionsClearRequested={props.clearSuggestions}
      onSuggestionsFetchRequested={props.fetchSuggestions}
      renderSuggestionsContainer={renderSuggestionsContainer}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      onSuggestionSelected={handleOnSuggestionSelected}
      inputProps={{
        classes,
        placeholder: 'Search place names',
        value: props.search.query,
        onChange: handleOnChange,
        onKeyDown: handleOnKeyDown
      }}
    />
  );
};

IntegrationAutosuggest.propTypes = {
  classes: PropTypes.object.isRequired,
  search: PropTypes.object.isRequired,
  updateQuery: PropTypes.func.isRequired,
  fetchSuggestions: PropTypes.func.isRequired,
  clearSuggestions: PropTypes.func.isRequired,
  fetchResults: PropTypes.func.isRequired,
};

export default withStyles(styles)(IntegrationAutosuggest);
