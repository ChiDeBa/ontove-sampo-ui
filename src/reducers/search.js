import {
  UPDATE_QUERY,
  UPDATE_DATASETS,
  UPDATE_SUGGESTIONS,
  CLEAR_SUGGESTIONS,
  UPDATE_RESULTS,
  CLEAR_RESULTS
} from '../actions';
import suggestions from './suggestions';
import results from './results';

export const INITIAL_STATE = {
  query: '',
  datasets: ['warsa_karelian_places', 'warsa_municipalities'],
  suggestions: [],
  results: []
};

const search = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case UPDATE_QUERY:
      return { ...state, query: action.query || '' };
    case UPDATE_DATASETS:
      return { ...state, datasets: action.datasets || [] };
    case CLEAR_SUGGESTIONS:
    case UPDATE_SUGGESTIONS:
      return { ...state, suggestions: suggestions(state.suggestions, action) };
    case CLEAR_RESULTS:
    case UPDATE_RESULTS:
      return {
        ...state,
        suggestions: [], 
        results: results(state.results, action)

      };
    default:
      return state;
  }
};

export default search;
