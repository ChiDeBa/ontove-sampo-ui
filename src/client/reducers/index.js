
import { combineReducers } from 'redux';
import {reducer as toastrReducer} from 'react-redux-toastr';
import {createResponsiveStateReducer} from 'redux-responsive';
import search from './search';
import error from './error';
import options from './options';
import map from './map';
import facet from './facet';

const reducer = combineReducers({
  options,
  search,
  map,
  facet,
  error,
  toastr: toastrReducer,
  browser: createResponsiveStateReducer({
    extraSmall: 500,
    small: 700,
    medium: 1000,
    large: 1400,
    extraLarge: 1600,
  }),
});

export default reducer;
