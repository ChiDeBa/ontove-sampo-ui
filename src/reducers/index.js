
import { combineReducers } from 'redux';
import {reducer as toastrReducer} from 'react-redux-toastr';
import search from './search';
import error from './error';
import options from './options';

const reducer = combineReducers({
  options,
  search,
  error,
  toastr: toastrReducer,
});

export default reducer;
