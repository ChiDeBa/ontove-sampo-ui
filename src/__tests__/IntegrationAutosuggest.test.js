import React from 'react';
import ReactDOM from 'react-dom';
import IntegrationAutosuggest from '../components/IntegrationAutosuggest';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { INITIAL_STATE } from '../reducers/search';

describe('IntegrationAutosuggest', () => {
  let updateQuery, fetchSuggestions, clearSuggestions;

  beforeEach(() => {
    updateQuery = jest.fn();
    fetchSuggestions = jest.fn();
    clearSuggestions = jest.fn();
  });

  test('Renders with initial state without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <IntegrationAutosuggest search={INITIAL_STATE} updateQuery={updateQuery}
        fetchSuggestions={fetchSuggestions} clearSuggestions={clearSuggestions} />,
      div
    );
  });

  test('Displays query string', () => {
    const newState = { ...INITIAL_STATE, query: 'kivennapa' };
    const component = renderer.create(
      <IntegrationAutosuggest search={newState} updateQuery={updateQuery}
        fetchSuggestions={fetchSuggestions} clearSuggestions={clearSuggestions} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Calls updateQuery and fetchSuggestions when user input changes', () => {
    const component = mount(
      <IntegrationAutosuggest search={INITIAL_STATE} updateQuery={updateQuery}
        fetchSuggestions={fetchSuggestions} clearSuggestions={clearSuggestions} />,
    );
    expect(fetchSuggestions.mock.calls.length).toBe(0);
    component.find('input').simulate('change', { target: { value: 'kivennapa' } });
    expect(updateQuery.mock.calls.length).toBe(1);
    expect(updateQuery.mock.calls[0][0]).toEqual('kivennapa');
    expect(fetchSuggestions.mock.calls.length).toBe(1);
  });

});
