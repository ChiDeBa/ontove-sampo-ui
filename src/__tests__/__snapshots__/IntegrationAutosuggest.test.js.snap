// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntegrationAutosuggest Displays query string 1`] = `
<div
  className="IntegrationAutosuggest-container-1"
>
  <div
    aria-describedby={undefined}
    className="MuiFormControl-root-5 MuiFormControl-fullWidth-8"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-expanded={false}
      aria-owns="react-autowhatever-1"
      className="MuiInput-root-9 MuiInput-fullWidth-16 MuiInput-formControl-10 MuiInput-underline-13"
      role="combobox"
      style={Object {}}
    >
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete="off"
        autoFocus={undefined}
        className="MuiInput-input-17"
        defaultValue={undefined}
        disabled={false}
        id={undefined}
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={undefined}
        placeholder="Search place names"
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="text"
        value="kivennapa"
      />
    </div>
  </div>
  <div
    className="MuiPaper-root-22 MuiPaper-elevation2-26"
    id="react-autowhatever-1"
    style={Object {}}
  />
</div>
`;

exports[`IntegrationAutosuggest Renders with initial state without crashing 1`] = `
<div
  className="IntegrationAutosuggest-container-1"
>
  <div
    aria-describedby={undefined}
    className="MuiFormControl-root-5 MuiFormControl-fullWidth-8"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-expanded={false}
      aria-owns="react-autowhatever-1"
      className="MuiInput-root-9 MuiInput-fullWidth-16 MuiInput-formControl-10 MuiInput-underline-13"
      role="combobox"
      style={Object {}}
    >
      <input
        aria-invalid={false}
        aria-required={false}
        autoComplete="off"
        autoFocus={undefined}
        className="MuiInput-input-17"
        defaultValue={undefined}
        disabled={false}
        id={undefined}
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={undefined}
        placeholder="Search place names"
        readOnly={undefined}
        required={undefined}
        rows={undefined}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="MuiPaper-root-22 MuiPaper-elevation2-26"
    id="react-autowhatever-1"
    style={Object {}}
  />
</div>
`;
